### PR TITLE
Change diff renamed files

### LINKED ISSUE #1038: Renamed files show diff as new files, don't show changes from old file or path to old file
**Describe the bug**
If I have staged changes that include a file rename, gitui will show the staged change with the "R" status, but it won't tell me what path it was renamed from, and the diff for the file shows the whole thing as new instead of diffing against the old file.

**To Reproduce**
1. Rename a file
2. Make some edits to the renamed file, for good measure
3. Open gitui
4. Stage the new file
5. Check the diff
6. Stage the removed file
7. Check the diff

**Expected behavior**
After staging just the new file, the diff should show as all new. After staging the removed file, the diff should indicate what it was renamed from (ideally the status would too), and show the diff from the new file against the old file.

**Actual behavior**
The diff after just staging the new file is correct. After staging the old file, the status shows "R" but the diff continues to act as though it's a brand new file instead of renamed file.

**Context (please complete the following information):**
 - macOS 12.0.1 (21A559)
 - 0.19.0
 - 1.56.1

### PR DESCRIPTION
**Pull request by Mifom**
This Pull Request closes #1038.

It changes the following:
- Add status change on rename
- Change diff title to represent rename
- Change diff to show only rewrites on rename

I followed the checklist:
- [X] I added and fixed current unittests.
- [x] I ran `make check` without errors
- [X] I tested the overall application
- [X] I added an appropriate item to the changelog

### CODE REVIEW

**Comment by extrawurst** [comment]
@Mifom Thanks for looking into this. I played a bit around with you branch. did not check the code yet but two observations jump out:

a) we still do not detect renames against the workdir, this should be easy to add:
Screenshot 2021-12-12 at 09 57 20

b) and then there is an issue staging hunks in the renamed&changed file:
Screenflick Movie 26

b) current hunk staging code is a bit wonky, so don't worry If thats the reason, then this might be better done in a separate PR

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## Unreleased
 
+### Added
+- Correct diff of renamed files [[@Mifom](https://github.com/Mifom)] ([#1035](https://github.com/extrawurst/gitui/issues/1038))
```
Comment (extrawurst) [inline CHANGELOG.md:11]: 1035 or 1038?

File: src/tabs/status.rs [hunk 2]
```diff
@@ -367,15 +367,20 @@ impl Status {
 		self.index.focus_select(is_stage);
 	}
 
-	pub fn selected_path(&self) -> Option<(String, bool)> {
+	pub fn selected_path(
+		&self,
+	) -> Option<(Option<String>, String, StatusItemType, bool)> {
+		// FIXME: replace with type
```
Comment (extrawurst) [inline src/tabs/status.rs:373]: I am using `//TODO` for these, lets keep this consistent

**Review by extrawurst — CHANGES_REQUESTED** [review]
lets add a dedicated unittest into asyncgit for the changed diff mechanic please

File: CHANGELOG.md (hunk 1 above)
Comment (Mifom) [reply CHANGELOG.md]: Fixed by https://github.com/extrawurst/gitui/pull/1040/commits/f487a6e7bab1e8e1aec8d4e7d2357200efb314e7

File: src/tabs/status.rs (hunk 2 above)
Comment (Mifom) [reply src/tabs/status.rs]: Fixed by https://github.com/extrawurst/gitui/pull/1040/commits/f487a6e7bab1e8e1aec8d4e7d2357200efb314e7

**Comment by Mifom** [comment]
I'll mark it ready when I deal with hunks

**Comment by Mifom** [comment]
For now hunks mostly done. But there are 2 issues:

1) Because destination file is untracked, diff in workdir is shown only for source file: 
Screenshot from 2021-12-17 16-58-49
git currently works the same way. Is this ok?

2) In staged renamed file is only one hunk - the whole file, so changes cannot be unstaged without unstaging full file:
Screenshot from 2021-12-17 17-18-37

**Comment by extrawurst** [comment]
> is this issue need to be fixed in this PR? Currently it's not supported by git2 and libgit2.

i guess then not :)
But then we should just go for regular file-level stagin/upstaging when enter is used on a hunk in a renamed file I stead of creating an error

**Comment by extrawurst** [comment]
Then this is ready ?

**Comment by Mifom** [comment]
> Then this is ready ?

Yes

**Comment by Mifom** [comment]
> > is this issue need to be fixed in this PR? Currently it's not supported by git2 and libgit2.
> 
> i guess then not :)
> But then we should just go for regular file-level stagin/upstaging when enter is used on a hunk in a renamed file I stead of creating an error

If I understand correctly this is how it works now

**Comment by extrawurst** [comment]
@Mifom so functionality wise this looks awesome to me. but I am not sure what I am missing, cause I cannot reproduce the single-hunk-issue:
Screenflick Movie 27

looks pretty good to me, with the only glitch that removing a hunk removes the rename-part of the change aswell. but I guess thats a mini thing we can followup later with. especially cause its a philosophical decision of what should take precendence I guess :)

what do you think?

File: asyncgit/src/sync/diff.rs [hunk 3]
```diff
@@ -189,21 +192,25 @@ pub(crate) fn get_diff_raw<'a>(
 		repo.diff_index_to_workdir(None, Some(&mut opt))?
 	};
 
+	diff.find_similar(Some(
```
Comment (extrawurst) [inline asyncgit/src/sync/diff.rs:195]: I think we should be able to get the same result by tweaking the diff options we pass in above

File: asyncgit/src/sync/diff.rs [hunk 4]
```diff
@@ -437,8 +444,14 @@ mod tests {
 
 		assert_eq!(get_statuses(repo_path), (1, 0));
 
-		let diff =
-			get_diff(repo_path, "foo/bar.txt", false, None).unwrap();
+		let diff = get_diff(
```
Comment (extrawurst) [inline asyncgit/src/sync/diff.rs:447]: is there also a test for the rename case?

File: asyncgit/src/sync/status.rs [hunk 5]
```diff
@@ -94,6 +98,21 @@ impl From<StatusType> for StatusShow {
...
 
+fn get_diff<'a>(
+	status_type: StatusType,
+	status_entry: &'a StatusEntry,
+) -> Option<DiffDelta<'a>> {
+	match status_type {
+		StatusType::WorkingDir => status_entry.index_to_workdir(),
+		StatusType::Stage => status_entry.head_to_index(),
+		StatusType::Both => {
+			status_entry // TODO: chain both Some(...) values
```
Comment (extrawurst) [inline asyncgit/src/sync/status.rs:109]: what about the TODO?

File: asyncgit/src/sync/status.rs [hunk 6]
```diff
@@ -150,16 +170,98 @@ pub fn get_status(
...
+		},
+	};
+	use std::{
+		fs::{self, File},
+		io::Write,
+		path::Path,
+	};
+
+	#[test]
+	fn test_rename_file() -> Result<()> {
```
Comment (extrawurst) [inline asyncgit/src/sync/status.rs:218]: not sure I understand this test: its not testing anything from this lib but just git2-rs, right?

we need a rename test for our diffing/status functions

File: src/tabs/status.rs [hunk 7]
```diff
@@ -367,15 +367,20 @@ impl Status {
 		self.index.focus_select(is_stage);
 	}
 
-	pub fn selected_path(&self) -> Option<(String, bool)> {
+	pub fn selected_path(
+		&self,
+	) -> Option<(Option<String>, String, StatusItemType, bool)> {
+		// TODO: replace with type
```
Comment (extrawurst) [inline src/tabs/status.rs:373]: lets do this

**Review by extrawurst — CHANGES_REQUESTED** [review]
just a few minor things left and a proper test for the rename case

**Comment by extrawurst** [comment]
I found another issue left:

Screenflick Movie 28

So discarding a renamed file item (now only a single entry in the status list) should discard fully, old-path and new-path. please also add a test for this case.

File: asyncgit/src/sync/status.rs (hunk 5 above)
Comment (Mifom) [reply asyncgit/src/sync/status.rs]: I've done that
https://github.com/extrawurst/gitui/pull/1040/commits/126140de30d84c3e4e55716b3bccd05365e0999f

File: asyncgit/src/sync/status.rs (hunk 6 above)
Comment (Mifom) [reply asyncgit/src/sync/status.rs:216]: Thanks for comment!
I fixed this
https://github.com/extrawurst/gitui/pull/1040/commits/7e45afda8bc8ee48d754fdc9f4dc39599c075047

File: src/tabs/status.rs (hunk 7 above)
Comment (Mifom) [reply src/tabs/status.rs]: Added DiffData
https://github.com/extrawurst/gitui/pull/1040/commits/28ac5addb76ff11695dffe7bb0952f69f36fbaca

File: asyncgit/src/sync/diff.rs (hunk 4 above)
Comment (Mifom) [reply asyncgit/src/sync/diff.rs:447]: Not sure that it test it, but git_diff now compare two files: old and new.

File: asyncgit/src/sync/diff.rs (hunk 3 above)
Comment (Mifom) [reply asyncgit/src/sync/diff.rs:195]: Sorry, but don't understand this comment.

File: asyncgit/src/sync/diff.rs (hunk 4 above)
Comment (extrawurst) [reply asyncgit/src/sync/diff.rs:447]: well I don't see that you added one, so we still need one for that case

**Comment by extrawurst** [comment]
@Mifom let me know when I should review it again

File: asyncgit/src/sync/diff.rs (hunk 3 above)
Comment (extrawurst) [reply asyncgit/src/sync/diff.rs:195]: all good

**Comment by extrawurst** [comment]
@Mifom how can I help you get this over the finish line?

**Comment by Mifom** [comment]
@extrawurst I just need some time. I think I'll do it before the end of the week.

**Comment by extrawurst** [comment]
@Mifom thanks for your continued effort, let me know when you think its ready

**Comment by extrawurst** [comment]
@Mifom ping. feel free to join the discord if you need some feedback somewhere

**Comment by extrawurst** [comment]
here is an issue when you want to discard a hunk in a renamed file:
Screenshot 2022-01-21 at 22 04 32

this will error-close with: 
```
Error: `hunk not found`
```

**Comment by extrawurst** [comment]
here is an issue with line-staging:
Screenshot 2022-01-21 at 22 07 16

leads to:
Screenshot 2022-01-21 at 22 07 54

as with above hunk error we can workaround this (if we don't wanna fix it now) by not allowing line based staging on renamed entries.

**Comment by extrawurst** [comment]
when there is a rename involving a directory change then folder based staging does not work in a single step:

https://user-images.githubusercontent.com/776816/150601785-6e129f43-c05e-4057-9083-86bcf6d68998.mp4

**Comment by Mifom** [comment]
The list of remaining changes:
 - [ ] test for rename file
 - [ ] test for discarding file
 - [ ] error with discarding hunk
 - [ ] error with line staging
 - [ ] issue with renaming directory

File: asyncgit/src/sync/diff.rs (hunk 4 above)
Comment (Mifom) [reply asyncgit/src/sync/diff.rs:454]: Test added at https://github.com/extrawurst/gitui/pull/1040/commits/1a199254edaaf3d3d1c4d801208bbcbae3aa9ea2

**Comment by Mifom** [comment]
@extrawurst sorry for this long period without work on this PR.

There are no good way for discarding hunks in renamed files at the moment, so this was disabled.

About line staging: it works at least for the changes i made. Maybe there are cases where it will not work, so i'll check that.

About renamed folder staging: at the moment when user mark folder as staged it passes to staging on the side of git2, so we haven't control this. That's should be reimplemented in app side to accurately remove them and not only add new files.

**Comment by extrawurst** [comment]
@Mifom can you bring the branch up to date with master, so it could in theory be merged?

**Comment by Mifom** [comment]
> @Mifom can you bring the branch up to date with master, so it could in theory be merged?

Now it's up to date

**Comment by Mifom** [comment]
@extrawurst the branch is now up to date
In the comment above I explain status of all issues with this PR. If you have ideas how to surpass them I can implement them or we can open new issue for them where they will be resolved.

**Comment by extrawurst** [comment]
Thanks for sticking with this! 🥳

**Comment by extrawurst** [comment]
I think I have to revert this. after the merge no `return` on any file in the file list works anymore to stage the whole file

**Comment by extrawurst** [comment]
ok damn I don't get the changes made to `index_add_remove` now that I try to debug this. you seem to `sync::reset_stage` anything that was just staged in the same function. sorry I did not catch this before the merge. I will revert the change. feel free to reopen the PR

**Comment by extrawurst** [comment]
@Mifom let me know if you need help figuring this out.

**Pull request merged**